internal/users/users.go

```diff
@@ -161,6 +165,116 @@ func computeGroupOperation(newGroupsInfo []GroupInfo, currentLocalGroups []strin
 	return groupsToAdd, groupsToRemove
 }
 
+// CleanUserFromSystemGroups removes the user from all system groups.
+func CleanUserFromSystemGroups(user string, args ...Option) (err error) {
+	defer decorate.OnError(&err, "could not clean user %q from system groups", user)
+
+	opts := defaultOptions
+	for _, arg := range args {
+		arg(&opts)
+	}
+
+	// Get the list of groups the user belong to
+	groups, err := existingLocalGroups(user, opts.groupPath)
+	if err != nil {
+		return err
+	}
+
+	for _, group := range groups {
+		args := opts.gpasswdCmd[1:]
+		args = append(args, "--delete", user, group)
+		if err := runGPasswd(opts.gpasswdCmd[0], args...); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// CleanSystemGroups removes all unexistent users from the system groups.
+//
+// In order to do that, it needs to query NSS for all existent users and then look at /etc/groups.
+func CleanSystemGroups(args ...Option) (err error) {
+	defer decorate.OnError(&err, "could not clean system groups completely")
+
+	opts := defaultOptions
+	for _, arg := range args {
+		arg(&opts)
+	}
+
+	// Get the list of users from NSS
+	// #nosec:G204 - the cmd arguments are only overridable in a controlled way during tests.
+	cmd := exec.Command(opts.getUsersCmd[0], opts.getUsersCmd[1:]...)
+	var out bytes.Buffer
+	cmd.Stdout = &out
+	if err := cmd.Run(); err != nil {
+		return fmt.Errorf("could not get list of users from NSS (%v): %s", err, out.String())
+	}
+
+	// Add the existingUsers to a map to speed up search
+	existingUsers := make(map[string]struct{})
+	for _, line := range strings.Split(out.String(), "\n") {
+		if line == "" {
+			continue
+		}
+		name, _, found := strings.Cut(line, ":")
+		if !found {
+			slog.Warn(fmt.Sprintf("could not parse NSS output: badly formatted entry %s", line))
+			continue
+		}
+		existingUsers[name] = struct{}{}
+	}
+
+	// Get the list of groups
+	f, err := os.Open(opts.groupPath)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	// Format of a line composing the group file is:
+	// group_name:password:group_id:user1,…,usern
+	var delOps [][]string
+	scanner := bufio.NewScanner(f)
+	for scanner.Scan() {
+		t := strings.TrimSpace(scanner.Text())
+		if t == "" {
+			continue
+		}
+		elems := strings.Split(t, ":")
+		if len(elems) != 4 {
+			e := fmt.Errorf("unexpected number of elements in group file on line (should have 4 separators): %q", t)
+			err = errors.Join(err, e)
+			continue
+		}
+
+		groupName := elems[0]
+		users := strings.Split(elems[3], ",")
+		for _, user := range users {
+			if _, ok := existingUsers[user]; ok {
+				continue
+			}
+
+			// User doesn't exist anymore, remove it from the group
+			args := opts.gpasswdCmd[1:]
+			delOps = append(delOps, append(args, "--delete", user, groupName))
+		}
+	}
+	if err := scanner.Err(); err != nil {
```

Comment:
I find it confusing that you are reusing "err" here compared to the one that has been accumulated earlier in errgroup, as indeed, this is another variable as in another scope, and then, the outer `err` will not be touched and accumated with the gpasswd cmd errors later before being returned.

I suggest that any malformed file is stopped right away. There is likely high chance that gpasswd will fail too anyway and we don’t want to touch corrupted files. You can still accumate the error, but you should consider it as `scanner.Err()` IMHO: stop this function ASAP.

Reply:
I can see that. I was using the same approach that we decided for the cache cleanup, i.e. "clean as much as possible and return an error if it couldn't clean one (or more) entries", but stopping at the first malformed entry is also fine (I think). Do you think it's better?

Also, I'm not sure whether accumulating the error is better than just logging when something goes wrong. Honestly, it feels like the behavior should be: 
1) if we want to error out on failure, return when the first one is found;
2) if we want to continue the operation, log if something goes wrong and continue;

Wdyt? (There can be some exceptions, but...) I feel like it's less confusing, no?

Reply:
> I can see that. I was using the same approach that we decided for the cache cleanup, i.e. "clean as much as possible and return an error if it couldn't clean one (or more) entries", but stopping at the first malformed entry is also fine (I think). Do you think it's better?
> 

I suggest that we still clean as much as possible if we are confident on the input. However, if we are not confident on the input file (malformed file), we before even starting cleaning.

This is actually what we do for other part of local group handling in `existingLocalGroups`, failing on first incorrect parsing.

> Also, I'm not sure whether accumulating the error is better than just logging when something goes wrong. Honestly, it feels like the behavior should be:
> 
> 1. if we want to error out on failure, return when the first one is found;
> 2. if we want to continue the operation, log if something goes wrong and continue;
> 
> Wdyt? (There can be some exceptions, but...) I feel like it's less confusing, no?

It depends. If the errors can be independent of each others, I feel like accumating is a bonus for the user, this avoids:

* -> there is an error, return it
* Ok, I have fixed it, let me return
* -> here for you another error sir
* oh, let me fix it, rerun
* -> oh btw, now, this is your next error…

I always find that, as an user, very annoying… Just tell me from the get go all the related things failing from the start if you can :)